L02 Exploring Deep Learning Tools: A No-Code Introduction to TensorFlow and Keras

Introduction

We read the introduction and understood that the main goal of this lab is to learn how deep learning and pre-trained models work using a no-code approach with Google Colab and VGG16.

In [ ]:
import tensorflow as tf
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input, decode_predictions
from tensorflow.keras.preprocessing import image
import numpy as np

model = VGG16(weights='imagenet')
model.summary()

553467096/553467096 ━━━━━━━━━━━━━━━━━━━━ 9s 0us/step


Model: "vgg16"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 224, 224, 3)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_conv1 (Conv2D)           │ (None, 224, 224, 64)   │         1,792 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_conv2 (Conv2D)           │ (None, 224, 224, 64)   │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_pool (MaxPooling2D)      │ (None, 112, 112, 64)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_conv1 (Conv2D)           │ (None, 112, 112, 128)  │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_conv2 (Conv2D)           │ (None, 112, 112, 128)  │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_pool (MaxPooling2D)      │ (None, 56, 56, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv1 (Conv2D)           │ (None, 56, 56, 256)    │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv2 (Conv2D)           │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv3 (Conv2D)           │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_pool (MaxPooling2D)      │ (None, 28, 28, 256)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv1 (Conv2D)           │ (None, 28, 28, 512)    │     1,180,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv2 (Conv2D)           │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv3 (Conv2D)           │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_pool (MaxPooling2D)      │ (None, 14, 14, 512)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv1 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv2 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv3 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_pool (MaxPooling2D)      │ (None, 7, 7, 512)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 25088)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ fc1 (Dense)                     │ (None, 4096)           │   102,764,544 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ fc2 (Dense)                     │ (None, 4096)           │    16,781,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ predictions (Dense)             │ (None, 1000)           │     4,097,000 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 138,357,544 (527.79 MB)

 Trainable params: 138,357,544 (527.79 MB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
from tensorflow.keras.utils import load_img, img_to_array

def prepare_image(img_path):
    img = load_img(img_path, target_size=(224, 224))
    x = img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    return x

In [ ]:
!wget -O elephant.jpg https://upload.wikimedia.org/wikipedia/commons/f/fc/African_Bush_Elephant.jpg

img_path = 'elephant.jpg'
image_data = prepare_image(img_path)
preds = model.predict(image_data)

results = decode_predictions(preds, top=3)[0]
for i, (imagenet_id, label, score) in enumerate(results):
    print(f"{i+1}. {label}: {score*100:.2f}%")

--2026-09-05 15:58:02--  https://upload.wikimedia.org/wikipedia/commons/f/fc/African_Bush_Elephant.jpg
Resolving upload.wikimedia.org (upload.wikimedia.org)... 198.35.26.240, 2620:0:863:ed1a::2:b
Connecting to upload.wikimedia.org (upload.wikimedia.org)|198.35.26.240|:443... connected.
HTTP request sent, awaiting response... 429 Too Many Requests (015d12c)
2026-09-05 15:58:02 ERROR 429: Too Many Requests (015d12c).



UnidentifiedImageError: cannot identify image file <_io.BytesIO object at 0x7bdaeed3e9d0>

In [ ]:
from google.colab import files
from tensorflow.keras.utils import load_img, img_to_array
import numpy as np

uploaded = files.upload()

for fn in uploaded.keys():
    img_path = fn
    image_data = prepare_image(img_path)
    preds = model.predict(image_data)

    results = decode_predictions(preds, top=3)[0]
    print(f"\n--- Results for {fn} ---")
    for i, (imagenet_id, label, score) in enumerate(results):
        print(f"{i+1}. {label}: {score*100:.2f}%")

Saving Classic Car.webp to Classic Car.webp
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step
35363/35363 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step

--- Results for Classic Car.webp ---
1. racer: 60.62%
2. sports_car: 38.92%
3. car_wheel: 0.20%


In [ ]:
import matplotlib.pyplot as plt

from tensorflow.keras.preprocessing.image import random_shift, random_rotation

img_to_test = load_img(fn, target_size=(224, 224))
x_orig = img_to_array(img_to_test)

x_modified = random_shift(x_orig, wrange=0.2, hrange=0.2, row_axis=0, col_axis=1, channel_axis=2)
x_input = np.expand_dims(x_modified, axis=0)
x_input = preprocess_input(x_input)

preds_mod = model.predict(x_input)
results_mod = decode_predictions(preds_mod, top=3)[0]

print(f"\n--- Results after modifying image (Step 5) ---")
for i, (imagenet_id, label, score) in enumerate(results_mod):
    print(f"{i+1}. {label}: {score*100:.2f}%")

TypeError: random_shift() got an unexpected keyword argument 'wrange'

In [ ]:
import numpy as np

x_modified = np.roll(x_orig, shift=15, axis=1)
x_input = np.expand_dims(x_modified, axis=0)
x_input = preprocess_input(x_input)

preds_mod = model.predict(x_input)
results_mod = decode_predictions(preds_mod, top=3)[0]

print(f"\n--- Results after modifying image (Step 5) ---")
for i, (imagenet_id, label, score) in enumerate(results_mod):
    print(f"{i+1}. {label}: {score*100:.2f}%")

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 904ms/step

--- Results after modifying image (Step 5) ---
1. racer: 61.37%
2. sports_car: 37.66%
3. cab: 0.43%


Conclusion and Discussion

We completed a 6-step deep learning lab in Google Colab using the pre-trained VGG16 model and ImageNet weights. This allowed us to test computer vision and image classification without writing code from scratch.


Step-by-Step Lab Process and Error Details:

Steps 1 and 2 (Model Setup): We imported TensorFlow and VGG16. Running model.summary() showed that the model has 138,357,544 parameters and takes 527.79 MB of memory, confirming it has the structure needed to extract image shapes and features.

Steps 3 and 4 (Data Upload and Prediction Error):

First Error: When we tried to download a default image (elephant.jpg) using wget, the server gave an HTTP 429 Too Many Requests error and failed.

The Fix: We used the interactive tool files.upload() to upload our own image of a car (Classic Car.webp) directly from our computer.

Prediction Result: The model successfully identified the car as racer (60.62%) and sports_car (38.92%), showing it works well for vehicle classes.

Step 5 (Pixel Shift and Testing Sensitivity):

Code Error: Trying to use Keras's random_shift function caused a TypeError because an argument (wrange) was not supported in this version.

The Fix: We replaced it with NumPy's shift function (np.roll(..., shift=15, axis=1)) to move the image pixels 15 pixels to the right.

Result Change: Moving the pixels slightly changed the prediction scores. The top prediction (racer) went up from 60.62% to 61.37%, and cab appeared in third place instead of car_wheel. This proves that deep learning models are sensitive to small changes in position and pixel alignment.

Final Reflection

This lab showed that pre-trained models like VGG16 can classify images instantly using transfer learning. However, it also proved that model outputs can shift when the input image pixels are moved or altered.

Feedback and Assessment

As a group, we think this lab was very practical and helpful to understand how pre-trained deep learning models work without writing complex code from scratch.

During the practice, we learned how to import the default model architecture called VGG16 using TensorFlow, how to handle image data preprocessing, and how to make successful predictions using our own classic car photo.

We also saw that AI models are very sensitive to spatial changes. When we modified and shifted the image pixels slightly, the confidence percentages and alternative labels changed, which clearly shows how visual alignment affects model behavior.